libmaven/mzkitchenutils.cpp

```diff
@@ -274,102 +277,168 @@ void MzKitchenProcessor::matchMetabolites(
 
-            FragmentationMatchScore s = library.scoreMatch(&(group.fragmentationPattern), params->ms2PpmTolr);
+            cout << "\n";
 
-            if (s.numMatches < params->ms2MinNumMatches) continue;
+        }
 
-            double normCosineScore = Fragment::normCosineScore(&library, &(group.fragmentationPattern), s.ranks);
-            s.dotProduct = normCosineScore;
+        Fragment library;
+        library.precursorMz = static_cast<double>(precMz);
+        library.mzs = compound->fragment_mzs;
+        library.intensity_array = compound->fragment_intensity;
+        library.fragment_labels = compound->fragment_labels;
 
-            scores.push_back(make_pair(compound, s));
+        //skip entries when the RT is required, and out of range
+        float rtDiff = abs(g->medianRt() - compound->expectedRt);
+        if (params->rtIsRequireRtMatch & (rtDiff > params->rtMatchTolerance)) {
+            continue;
         }
 
-        //Issue 559: print empty groups
+        FragmentationMatchScore s = library.scoreMatch(&(g->fragmentationPattern), params->ms2PpmTolr);
+
+        double normCosineScore = Fragment::normCosineScore(&library, &(g->fragmentationPattern), s.ranks);
+        s.dotProduct = normCosineScore;
+
+        //debugging
         if (debug) {
-            cout << group.meanMz << "@" << group.medianRt() << " #ms2s=" << group.ms2EventCount << " :" << endl;
+            cout << "Candidate Score: " << compound->id << ":\n";
+            cout << "numMatches= " << s.numMatches
+                 << ", hyperGeometricScore= " << s.hypergeomScore
+                 << ", cosineScore= " << s.dotProduct
+                 << " VS params->ms2MinNumMatches = " << params->ms2MinNumMatches
+                 << " " << (s.numMatches >= params->ms2MinNumMatches ? "yes" : "no")
+                 << "\n\n\n";
         }
 
-        if (scores.empty()) continue;
+        if (s.numMatches < params->ms2MinNumMatches) continue;
 
-        group.compounds = scores;
 
-        float maxScore = -1.0f;
-        pair<Compound*, FragmentationMatchScore> bestPair;
-        for (auto score : scores) {
-            if (score.second.dotProduct > maxScore) {
-                bestPair = score;
-                maxScore = score.second.dotProduct;
-            }
-        }
+        scores.push_back(make_pair(ion, s));
+    }
+
+    //Issue 559: print empty groups
+    if (debug) {
+        cout << g->meanMz << "@" << g->medianRt() << " #ms2s=" << g->ms2EventCount << " :" << endl;
+    }
+
+    if (scores.empty()) return;
 
-        if (maxScore > -1.0f) {
-            group.compound = bestPair.first;
-            group.fragMatchScore = bestPair.second;
-            group.fragMatchScore.mergedScore = bestPair.second.dotProduct;
+    g->compounds = scores;
+
+    float maxScore = -1.0f;
+    pair<CompoundIon, FragmentationMatchScore> bestPair;
+    for (auto score : scores) {
+        if (score.second.dotProduct > maxScore) {
+            bestPair = score;
```

Comment:
Currently, mzkitchen is relying entirely on the dotProduct score, but if this changes in the future, the changes will be reflected both in the MAVEN app and in the shiny app.